pkg/cloudprovider/cloudprovider.go

```diff
@@ -43,32 +51,144 @@ type CloudProvider struct {
 	recorder   events.Recorder
 
 	instanceTypeProvider instancetype.Provider
+	instanceProvider     instance.Provider
 }
 
 func New(kubeClient client.Client,
 	recorder events.Recorder,
 	instanceTypeProvider instancetype.Provider,
+	instanceProvider instance.Provider,
 ) *CloudProvider {
 	return &CloudProvider{
 		kubeClient:           kubeClient,
 		recorder:             recorder,
 		instanceTypeProvider: instanceTypeProvider,
+		instanceProvider:     instanceProvider,
 	}
 }
 
 // Create a NodeClaim given the constraints.
 func (c *CloudProvider) Create(ctx context.Context, nodeClaim *karpv1.NodeClaim) (*karpv1.NodeClaim, error) {
-	return nil, fmt.Errorf("not implemented")
+	log.FromContext(ctx).Info("creating nodes", "nodeClaim", nodeClaim)
+	nodeClass, err := c.resolveNodeClassFromNodeClaim(ctx, nodeClaim)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.recorder.Publish(cloudproviderevents.NodeClaimFailedToResolveNodeClass(nodeClaim))
+		}
+		return nil, cloudprovider.NewInsufficientCapacityError(fmt.Errorf("resolving node class, %w", err))
+	}
+
+	nodeClassReady := nodeClass.StatusConditions().Get(status.ConditionReady)
+	if nodeClassReady.IsFalse() {
+		return nil, cloudprovider.NewNodeClassNotReadyError(stderrors.New(nodeClassReady.Message))
+	}
+	if nodeClassReady.IsUnknown() {
+		return nil, cloudprovider.NewCreateError(fmt.Errorf("resolving NodeClass readiness, NodeClass is in Ready=Unknown, %s", nodeClassReady.Message), "NodeClassReadinessUnknown", "NodeClass is in Ready=Unknown")
+	}
+
+	instanceTypes, err := c.resolveInstanceTypes(ctx, nodeClass)
+	if err != nil {
+		return nil, cloudprovider.NewCreateError(fmt.Errorf("resolving instance types, %w", err), "InstanceTypeResolutionFailed", "Error resolving instance types")
+	}
+	if len(instanceTypes) == 0 {
+		return nil, cloudprovider.NewInsufficientCapacityError(fmt.Errorf("all requested instance types were unavailable during launch"))
+	}
+
+	instance, err := c.instanceProvider.Create(ctx, nodeClass, nodeClaim, instanceTypes)
+	if err != nil {
+		return nil, fmt.Errorf("creating instance, %w", err)
+	}
+
+	instanceType, _ := lo.Find(instanceTypes, func(i *cloudprovider.InstanceType) bool {
+		return i.Name == instance.Type
+	})
+
+	nc := c.instanceToNodeClaim(instance, instanceType)
+	nc.Annotations = lo.Assign(nc.Annotations, map[string]string{
+		v1alpha1.AnnotationGCENodeClassHash:        nodeClass.Hash(),
+		v1alpha1.AnnotationGCENodeClassHashVersion: v1alpha1.GCENodeClassHashVersion,
+	})
+	return nc, nil
 }
 
 func (c *CloudProvider) List(ctx context.Context) ([]*karpv1.NodeClaim, error) {
-	return nil, fmt.Errorf("not implemented")
+	log.FromContext(ctx).Info("listing instances from instance provider")
+	instances, err := c.instanceProvider.List(ctx)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "failed to list instances")
+		return nil, fmt.Errorf("listing instances, %w", err)
+	}
 
+	var nodeClaims []*karpv1.NodeClaim
+	for i := range instances {
```

Comment:
Wouldn't it be better here to do like:
```
for _, instance := range instances {
```

And use `instance` instead of `instances[i]` inside the loop

Reply:
I agree with it, and I have update this operation.